enclave-runtime/src/lib.rs

```diff
@@ -491,42 +491,36 @@ where
 	PB: BlockT<Hash = H256>,
 	NumberFor<PB>: BlockNumberOps,
 {
-	// we acquire lock explicitly (variable binding), since '_' will drop the lock after the statement
-	// see https://medium.com/codechain/rust-underscore-does-not-bind-fec6a18115a8
-	let _light_client_lock = EnclaveLock::write_light_client_db()?;
+	let validator_access = ValidatorAccessor::<PB>::default();
+	let genesis_hash = validator_access.execute_on_validator(|v| v.genesis_hash(v.num_relays()))?;
 
-	let mut validator = LightClientSeal::<PB>::unseal()?;
 	let signer = Ed25519Seal::unseal()?;
 	let stf_executor = StfExecutor::new(Arc::new(OcallApi), Arc::new(GlobalFileStateHandler));
-	let genesis_hash = validator.genesis_hash(validator.num_relays())?;
 	let extrinsics_factory =
 		ExtrinsicsFactory::new(genesis_hash, signer, GLOBAL_NONCE_CACHE.clone());
 
 	sync_blocks_on_light_client(
 		blocks_to_sync,
-		&mut validator,
+		&validator_access,
 		&extrinsics_factory,
 		&OcallApi,
 		&stf_executor,
 	)?;
 
-	// store updated state in light client in case we fail afterwards.
-	LightClientSeal::seal(validator)?;
-
 	Ok(())
 }
 
-fn sync_blocks_on_light_client<PB, V, OCallApi, StfExecutor, ExtrinsicsFactory>(
+fn sync_blocks_on_light_client<PB, ValidatorAccessor, OCallApi, StfExecutor, ExtrinsicsFactory>(
 	blocks_to_sync: Vec<SignedBlockG<PB>>,
-	validator: &mut V,
+	validator_access: &ValidatorAccessor,
 	extrinsics_factory: &ExtrinsicsFactory,
 	on_chain_ocall_api: &OCallApi,
 	stf_executor: &StfExecutor,
 ) -> Result<()>
```

Comment:
Instead of passing in a `Validator` directly (which would require holding a lock the entire time this validator is valid), we pass in the accessor. Locking then only has to be done when we actually use the validator, on a per-operation basis.